docs/architecture/adr-038-state-listening.md

```diff
@@ -305,126 +412,198 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
+	// that are written in Go.
+	Impl baseapp.ABCIListener
+}
 
-	// Init is called once when the Plugin is being loaded
-	// The plugin is passed the AppOptions for configuration
-	// A plugin will not necessarily have a functional Init
-	Init(env serverTypes.AppOptions) error
+func (p *ListenerGRPCPlugin) GRPCServer(_ *plugin.GRPCBroker, s *grpc.Server) error {
+	RegisterABCIListenerServiceServer(s, &GRPCServer{Impl: p.Impl})
+	return nil
+}
 
-	// Closer interface for shutting down the plugin process
-	io.Closer
+func (p *ListenerGRPCPlugin) GRPCClient(
+	_ context.Context,
+	_ *plugin.GRPCBroker,
+	c *grpc.ClientConn,
+) (interface{}, error) {
+	return &GRPCClient{client: NewABCIListenerServiceClient(c)}, nil
 }
 ```
 
-The `Name` method returns a plugin's name.
-The `Version` method returns a plugin's version.
-The `Init` method initializes a plugin with the provided `AppOptions`.
-The io.Closer is used to shut down the plugin service.
+The `plugin.Plugin` interface has two methods `Client` and `Server`. For our GRPC service these are `GRPCClient` and `GRPCServer`
+The `Impl` field holds the concrete implementation of our `baseapp.ABCIListener` interface written in Go.
+Note: this is only used for plugins that are written in Go.
 
-For the purposes of this ADR we introduce a single kind of plugin- a state streaming plugin.
-We will define a `StateStreamingPlugin` interface which extends the above `Plugin` interface to support a state streaming service.
+The `StreamingService` protocol:
 
-```go
-// StateStreamingPlugin interface for plugins that load a baseapp.StreamingService onto a baseapp.BaseApp
-type StateStreamingPlugin interface {
-	// Register configures and registers the plugin streaming service with the BaseApp
-	Register(bApp *baseapp.BaseApp, marshaller codec.BinaryCodec, keys map[string]*types.KVStoreKey) error
+```protobuf
+syntax = "proto3";
+
+package cosmos.sdk.grpc.abci.v1;
+
+option go_package           = "github.com/cosmos/cosmos-sdk/streaming/plugins/abci/grpc_abci_v1";
+option java_multiple_files  = true;
+option java_outer_classname = "AbciListenerProto";
+option java_package         = "network.cosmos.sdk.grpc.abci.v1";
+
+// PutRequest is used for storing ABCI request and response
+// and Store KV data for streaming to external grpc service.
+message PutRequest {
+  int64 block_height      = 1;
+  bytes req               = 2;
+  bytes res               = 3;
+  bytes store_kv_pair     = 4;
+  int64 store_kv_pair_idx = 5;
+  int64 tx_idx            = 6;
+}
 
-	// Start starts the background streaming process of the plugin streaming service
-	Start(wg *sync.WaitGroup) error
+message Empty {}
 
-	// Plugin is the base Plugin interface
-	Plugin
+service ABCIListenerService {
+  rpc ListenBeginBlock(PutRequest) returns (Empty);
+  rpc ListenEndBlock(PutRequest) returns (Empty);
+  rpc ListenDeliverTx(PutRequest) returns (Empty);
+  rpc ListenStoreKVPair(PutRequest) returns (Empty);
```

Comment:
What is the motivation to avoid the ABCI request types defined in tendermint here and instead use a generic `PutRequest`?